### PR TITLE
Battle service logic

### PR DESCRIPTION
**Pull request by Mgarchuk**

### CODE REVIEW

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 1]
```diff
@@ -1,43 +1,81 @@
...
+                .map(gameMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @GetMapping
-    public List<GameDto> getActiveGames() {
-        return null;
+    @GetMapping("/find-game")
+    public GameDto findGame(@Valid @RequestBody SettingsDto settingsDto) {
+        SettingsEntity settingsEntity = settingsMapper.toEntity(settingsDto);
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: what was the intent to find a game by SettingsDto?
you may use find by id which will return single possible game
or you can use findBY SettingsDto which may return list of games which fit to specific settings

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 2]
```diff
@@ -1,43 +1,81 @@
...
     @PostMapping("/create")
-    public GameDto createGame(@RequestBody GameDto gameDto) {
-        return null;
+    public GameDto createGame(@Valid @RequestBody GameDto gameDto) {
+        GameEntity gameEntity = gameMapper.toEntity(gameDto);
+        gameEntity = gameService.create(gameEntity, UUID.randomUUID());
+        return gameMapper.toDTO(gameEntity);
     }
 
+    //ToDo: finish this method
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: lets avoid "link". it's not clear what this for.
you can use gameId to join/get/update specific game.

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 3]
```diff
@@ -1,43 +1,81 @@
...
+        gameEntity = gameService.joinGame(gameEntity, userId);
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @PutMapping("/leave")
-    public GameDto leaveGame(@RequestBody GameDto gameDto) {
-        return null;
+    //ToDo: change userId as parameter to userId from authorization
+    @PutMapping("/leave/{userId}")
+    public GameDto leaveGame(@Valid @RequestBody GameDto gameDto, @PathVariable UUID userId) {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: you don't need to put game dto here.
you can use
/game/{gameId}/leave
and get user id from authorization context(you don't have authorization so far, but it's fine)
`
UserAuthentication authentication = (UserAuthentication) SecurityContextHolder.getContext().getAuthentication();`

File: battle-service/src/main/java/com/company/battle/controllers/MoveController.java [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
+        List<MoveEntity> moveEntities = moveService.getByGameId(gameId);
+        return moveEntities.stream()
+                .map(moveMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @PostMapping
-    public MoveDto addMove(@RequestBody MoveDto moveDto) {
-        return null;
+    @PostMapping("/create")
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/MoveController.java]: this method is not following REST principals.
i would say you should use
POST /game/{gameId}/move

File: battle-service/src/main/java/com/company/battle/repositories/GameRepository.java [hunk 5]
```diff
@@ -1,14 +1,36 @@
...
+    //ToDo: add sort by date and limit 1
+    @Query(value = "SELECT * FROM game g WHERE g.square_size = :squareSize and g.lines_count_for_win = :linesCount and g.move_time_limit = :timeLimit and g.status = 'PENDING' " +
+            "and g.visibility = 'PUBLIC' and g.o_player_id is not null",
+            nativeQuery = true)
+    Optional<GameEntity> findGameForXPlayer(@Param("squareSize") int squareSize, @Param("linesCount") int linesCountForWin, @Param("timeLimit") int moveTimeLimit);
+
+    @Query(value = "SELECT * FROM game g WHERE g.square_size = :squareSize and g.lines_count_for_win = :linesCount and g.move_time_limit = :timeLimit and g.status = 'PENDING' " +
+            "and g.visibility = 'PUBLIC' and g.x_player_id is not null",
+            nativeQuery = true)
+    Optional<GameEntity> findGameForOPlayer(@Param("squareSize") int squareSize, @Param("linesCount") int linesCountForWin, @Param("timeLimit") int moveTimeLimit);
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/repositories/GameRepository.java]: it's too hard to follow.
you may use JPA naming to simplify queues https://docs.spring.io/spring-data/jpa/docs/current/reference/html/,
not very clear why you need squareSize,linesCountForWin,moveTimeLimit in all of these queries
findGamebyUserXorByUserO(userId) may return lis of all games in which player particiapated(completed and in progress. you may add additional parameters to filter such games by status.


this is related to all other methods in repositories.
try to segregate queries.

File: battle-service/src/main/java/com/company/battle/services/GameService.java [hunk 6]
```diff
@@ -1,27 +1,111 @@
...
 
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
 import java.util.UUID;
 
 @Service
+@RequiredArgsConstructor
 public class GameService {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/services/GameService.java:32]: there are some comments in upstream classes like controllers and dtos which will lead to changes in a service. will review after next update.

**Review by Volad — CHANGES_REQUESTED** [review]
some comments

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 1 above)
Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: This is a search for a created game with the same settings. That is, a person has entered the settings that he needs. For example: squareSize = 3, linesCountForWin = 3 and other. And then the search happens. If no one has created such a game yet, then null will return and instead of searching there will be a request to create a game. And if there are such games, then the one that was created the most time ago is taken

File: battle-service/src/main/java/com/company/battle/repositories/GameRepository.java (hunk 5 above)
Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/repositories/GameRepository.java]: These methods are needed in order to find a game with the Pending status and with the same parameters that the player wants to play with. Perhaps you misunderstood the purpose of these methods. This methods search opponents. If you use named methods, the method name will be too long (with setting parameters)

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 2 above)
Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: Links are needed to invite a friend to the game. For example, send it to private messages on any social network, then a person can click on the link and join the game. Should I remove the link or leave it?

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 3 above)
Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: I changed gameDto with gameId. As for the userId, I will change it to the proposed option later, when the authorization  works. For this, I left a comment above the method. For now, I'm just mocking userId

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 1 above)
Comment (Volad) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: at least there might be more than 1 game fit to the settings you have listed, while you are returning single Game Dto.
you should use appropriate naming to get the  most "recently" created game which fit to the search criteria.
or change strategy to search games.

i would create separate SearchGameRequestDto where enumerate all possible search criteria ,and respond with List<GameDto> to allow user to choose from the list of options, this search requestdto may be handled in different ways, for examp,e using jpa criteria builder, and not usage of prepared queue.

for now i would limit functionality by returning just list of games in PENDING status.

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 2 above)
Comment (Volad) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: yeah, lets probably remove, you can use id embedded to the link, and extract id in FE application. it will be enough for MVP

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java (hunk 1 above)
Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: Okay, I can rename the method. But the idea of this method is that when the user starts searching for a game by specific parameters (by parameters in the settings), you need to return any suitable game to him. And I don't think it's right to search for the most recent game because older games will hang. Now, on the contrary, I return one of the oldest games according to these parameters. And I don't think it's appropriate to return all games in pending status, because it makes no sense to return a 4*4 game if the user is looking for a 3*3 game

Comment (Volad) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: i don't understand what are you talking about:)
what i want to say is that settingsdto is something which consist of predefined configuration, so as long as there is predefined configuration user will have limited number search result (it may 1 , or 1000) depending on how many people use the same settings to create theirs games.
it doesn't make sense to return "ANY" single game from api standpoint, due to it makes your api very limited,and not extandable. what if tomorrow requirement will tell you "return 2 games which fit to the search criteria"?
usually you design api for consumers, and consumers able to control own behavior based on your api.
if you need to return just 1 game - yor method may still return "list" but there should be "limit" request param which will control number of max results in response. so consumer may say "return me list of N games which best fit to my search criteria", today N may be equal to 1, and tomorrow it may equal to 100. but in this case only consumers will require to adopt businesss logic, and not you, as backend developer.

File: battle-service/src/main/java/com/company/battle/repositories/GameRepository.java (hunk 5 above)
Comment (Volad) [reply battle-service/src/main/java/com/company/battle/repositories/GameRepository.java]: it's better to use jpa criteria builder instead of keep sql query hardcoded. i would suggest to start from small pieces.

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 7]
```diff
@@ -1,43 +1,74 @@
...
+        List<GameEntity> gameEntities = gameService.getPublicGames();
+        return gameEntities.stream()
+                .map(gameMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @GetMapping
-    public List<GameDto> getActiveGames() {
-        return null;
+    @GetMapping("/find-game")
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: /game/search

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 8]
```diff
@@ -1,43 +1,74 @@
...
+        gameEntity = gameService.create(gameEntity, UUID.randomUUID());
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @PutMapping("/join/{link}")
-    public GameDto joinGame(@PathVariable String link) {
-        return null;
+    //ToDo: Allow connection to only 1 game
+    @PutMapping("/join/{id}")
+    public GameDto joinGame(@PathVariable UUID id, @PathVariable UUID userId) {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: `  public GameDto joinGame(@PathVariable UUID id) {`
user id should be exteracted from security context. even if you don't have it, you can use fake value so far

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 9]
```diff
@@ -1,43 +1,74 @@
...
+        gameEntity = gameService.joinGame(gameEntity, userId);
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @PutMapping("/leave")
-    public GameDto leaveGame(@RequestBody GameDto gameDto) {
-        return null;
+    //ToDo: change userId as parameter to userId from authorization
+    @PutMapping("/game/{gameId}/user/{userId}/leave")
+    public GameDto leaveGame(@PathVariable UUID gameId, @PathVariable UUID userId) {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: "/{gameId}/leave")

File: battle-service/src/main/java/com/company/battle/controllers/MoveController.java [hunk 10]
```diff
@@ -1,22 +1,41 @@
...
+@RequiredArgsConstructor
 public class MoveController {
 
+    @Autowired
+    private final MoveService moveService;
+
+    @Autowired
+    private final MoveMapper moveMapper;
+
     @GetMapping("/{gameId}")
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/MoveController.java]: its a bit odd.
your request will look like 
api/v1/move/gameid

you can make this request part of game controller, or change uri
api/v1/game/gameid/moves

File: battle-service/src/main/java/com/company/battle/controllers/MoveController.java [hunk 11]
```diff
@@ -1,22 +1,41 @@
...
+        List<MoveEntity> moveEntities = moveService.getByGameId(gameId);
+        return moveEntities.stream()
+                .map(moveMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @PostMapping
-    public MoveDto addMove(@RequestBody MoveDto moveDto) {
-        return null;
+    @PostMapping("/game/{gameId}/move")
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/MoveController.java]: not really like path here
move/game/gameid/move

i would make it part of game controller or rename path
POST game/gameid/move

File: battle-service/src/main/java/com/company/battle/services/GameService.java [hunk 12]
```diff
@@ -1,27 +1,141 @@
...
+    }
+
+    public List<GameEntity> findGame(SearchGameRequestDto requestDto) {
+
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<GameEntity> criteriaQuery = criteriaBuilder.createQuery(GameEntity.class);
+        Root<GameEntity> root = criteriaQuery.from(GameEntity.class);
+
+        Predicate[] predicates = new Predicate[6];
+        predicates[0] = criteriaBuilder.equal(root.get("settings").get("squareSize"), requestDto.getSquareSize());
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/services/GameService.java]: what if i user didn't pass square size or any of other parameters.

i.e. i want to find games only with linesCountToWin eq 3 
?

i think you should additionally filter by non empty values in incoming parameters.

File: battle-service/src/main/java/com/company/battle/services/MoveService.java [hunk 13]
```diff
@@ -1,19 +1,65 @@
...
-        return null;
+    @Autowired
+    private final MoveRepository moveRepository;
+
+    @Autowired
+    private final GameRepository gameRepository;
+
+    public MoveEntity add(MoveEntity moveEntity, UUID gameId) {
+
+        if (gameId != moveEntity.getGame().getId()) {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/services/MoveService.java]: you may ignore/override moveentity.gameId here . and remove this check.

Comment (Volad) [reply battle-service/src/main/java/com/company/battle/services/MoveService.java]: you should select game by gameId and check status .you should not work with nested entity (MoveEntity#game) here. 
due to it will be unclear for upstream classes that they need to construct whole entity before calling this method.

File: battle-service/src/main/java/com/company/battle/utils/GameUtils.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+import com.company.common.models.GameEntity;
+import com.company.common.models.MoveEntity;
+import com.company.common.models.UserEntity;
+import com.company.common.models.enums.GameStatus;
+
+import java.util.Map;
+import java.util.UUID;
+
+public class GameUtils {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/utils/GameUtils.java]: try to split this class to multiple utils/services.

GameValidationService
GameWinnerService

File: battle-service/src/main/java/com/company/battle/utils/GameUtils.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+        if (movesMap.get(new Coordinate(x, y)) != null) {
+            return false;
+        }
+
+        return (movesMap.size() % 2 == 0 && move.getUser().getId().equals(move.getGame().getSettings().getXPlayerId()))
+                || (movesMap.size() % 2 == 1 && move.getUser().getId().equals(move.getGame().getSettings().getOPlayerId()));
+    }
+
+    public static boolean isValidGame(GameEntity game) {
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/utils/GameUtils.java]: not clear `Valid` for what exactly?

Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/utils/GameUtils.java]: Renamed. So okay?

File: battle-service/src/main/java/com/company/battle/controllers/GameController.java [hunk 16]
```diff
@@ -1,43 +1,74 @@
...
+                .map(gameMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @GetMapping
-    public List<GameDto> getActiveGames() {
-        return null;
+    @GetMapping("/search")
+    public List<GameDto> searchGames(@Valid @RequestBody SearchGameRequestDto requestDto) {
+        List<GameEntity> gameEntities = gameService.findGame(requestDto);
```
Comment (Volad) [inline battle-service/src/main/java/com/company/battle/controllers/GameController.java]: i would introduce GameSearchFacade which will accept SearchGameRequestDto and construct CriteriaBuilder, after that it will call 
GameService.findbGame(CriteriaBuilder searchCriteria)

Comment (Mgarchuk) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: I added specifications. Do I no longer need to make a GameSearchFacade?

Comment (Volad) [reply battle-service/src/main/java/com/company/battle/controllers/GameController.java]: not for now.

**Review by Volad — APPROVED** [review]
